File: lib/bloomy/operations/headlines.rb
# frozen_string_literal: true

require "bloomy/utils/get_user_id"

module Bloomy
  class Headline
    include Bloomy::Utilities::UserIdUtility
    # Initializes a new headline instance
    #
    # @param conn [Object] the connection object to interact with the API
    def initialize(conn)
      @conn = conn
    end

    # Creates a new headline
    #
    # @param meeting_id [Integer] the ID of the meeting
    # @param title [String] the title of the headline
    # @param owner_id [Integer] the ID of the headline owner
    # @param notes [String] additional notes for the headline
    # @return [Hash] containing id, title, owner_details, and notes_url
    def create(meeting_id:, title:, owner_id: user_id, notes: nil)
      response = @conn.post("/api/v1/L10/#{meeting_id}/headlines",
        {title: title, ownerId: owner_id, notes: notes}.to_json)
      raise "Failed to create headline" unless response.status == 200

      {
        id: response.body["Id"],
        title: response.body["Name"],
        owner_details: {id: response.body["OwnerId"]},
        notes_url: response.body["DetailsUrl"]
      }
    end

    # Updates a headline
    #
    # @param headline_id [Integer] the ID of the headline to update
    # @param title [String] the new title of the headline
    # @return [Boolean] true if update was successful
    def update(headline_id:, title:)
      response = @conn.put("/api/v1/headline/#{headline_id}", {title: title}.to_json)
      raise "Failed to update headline" unless response.status == 200
      true
    end

    # Get headline details
    #
    # @param headline_id [Integer] the ID of the headline
    # @return [Hash] containing id, title, notes_url, meeting_details,
    #                owner_details, archived, created_at, and closed_at
    def details(headline_id)
      response = @conn.get("/api/v1/headline/#{headline_id}?Include_Origin=true")
      raise "Failed to get headline details" unless response.status == 200

      {
        id: response.body["Id"],
        title: response.body["Name"],
        notes_url: response.body["DetailsUrl"],
        meeting_details: {
          id: response.body["OriginId"],
          title: response.body["Origin"]
        },
        owner_details: {
          id: response.body["Owner"]["Id"],
          name: response.body["Owner"]["Name"]
        },
        archived: response.body["Archived"],
        created_at: response.body["CreateTime"],
        closed_at: response.body["CloseTime"]
      }
    end

    # Get headlines for a user or a meeting.
    #
    # @param user_id [Integer, nil] the ID of the user (defaults to initialized user_id)
    # @param meeting_id [Integer, nil] the ID of the meeting
    # @raise [ArgumentError] if both `user_id` and `meeting_id` are provided
    # @return [Array<Hash>] a list of headlines containing:
    #   - id
    #   - title
    #   - meeting_details
    #   - owner_details
    #   - archived
    #   - created_at
    #   - closed_at
    # @example
    #   client.headline.list
    #   #=> [
    #     {
    #       id: 1,
    #       title: "Headline Title",
    #       meeting_details: { id: 1, title: "Team Meeting" },
    #       owner_details: { id: 1, name: "John Doe" },
    #       archived: false,
    #       created_at: "2023-01-01",
    #       closed_at: nil
    #     }
    #   ]
    def list(user_id: nil, meeting_id: nil)
      raise ArgumentError, "Please provide either `user_id` or `meeting_id`, not both." if user_id && meeting_id

      if meeting_id
        response = @conn.get("/api/v1/l10/#{meeting_id}/headlines")
      else
        user_id ||= self.user_id
        response = @conn.get("/api/v1/headline/users/#{user_id}")
      end

      raise "Failed to list headlines" unless response.success?

      response.body.map do |headline|
        {
          id: headline["Id"],
          title: headline["Name"],
          meeting_details: {
            id: headline["OriginId"],
            title: headline["Origin"]
          },
          owner_details: {
            id: headline["Owner"]["Id"],
            name: headline["Owner"]["Name"]
          },
          archived: headline["Archived"],
          created_at: headline["CreateTime"],
          closed_at: headline["CloseTime"]
        }
      end
    end

    # Deletes a headline
    #
    # @param headline_id [Integer] the ID of the headline to delete
    # @return [Boolean] true if the deletion was successful
    def delete(headline_id)
      response = @conn.delete("/api/v1/headline/#{headline_id}")
      response.success?
    end
  end
end

File: src/bloomy/operations/headlines.py
"""Headline operations for the Bloomy SDK."""

from __future__ import annotations

import builtins

from ..models import (
    HeadlineDetails,
    HeadlineInfo,
    HeadlineListItem,
    OwnerDetails,
)
from ..utils.base_operations import BaseOperations
from .mixins.headlines_transform import HeadlineOperationsMixin


class HeadlineOperations(BaseOperations, HeadlineOperationsMixin):
    """Class to handle all operations related to headlines."""

    def create(
        self,
        meeting_id: int,
        title: str,
        owner_id: int | None = None,
        notes: str | None = None,
    ) -> HeadlineInfo:
        """Create a new headline.

        Args:
            meeting_id: The ID of the meeting
            title: The title of the headline
            owner_id: The ID of the headline owner (default: current user ID)
            notes: Additional notes for the headline

        Returns:
            A HeadlineInfo model instance containing id, title, owner_details,
            and notes_url

        """
        if owner_id is None:
            owner_id = self.user_id

        payload = {"title": title, "ownerId": owner_id}
        if notes is not None:
            payload["notes"] = notes

        response = self._client.post(f"L10/{meeting_id}/headlines", json=payload)
        response.raise_for_status()
        data = response.json()

        return HeadlineInfo(
            id=data["Id"],
            title=data["Name"],
            owner_details=OwnerDetails(id=owner_id, name=None),
            notes_url=data.get("DetailsUrl", ""),
        )

    def update(self, headline_id: int, title: str) -> None:
        """Update a headline.

        Args:
            headline_id: The ID of the headline to update
            title: The new title of the headline

        """
        payload = {"title": title}
        response = self._client.put(f"headline/{headline_id}", json=payload)
        response.raise_for_status()

    def details(self, headline_id: int) -> HeadlineDetails:
        """Get headline details.

        Args:
            headline_id: The ID of the headline

        Returns:
            A HeadlineDetails model instance containing id, title, notes_url,
            meeting_details, owner_details, archived, created_at, and closed_at

        """
        response = self._client.get(
            f"headline/{headline_id}", params={"Include_Origin": "true"}
        )
        response.raise_for_status()
        data = response.json()

        return self._transform_headline_details(data)

    def list(
        self, user_id: int | None = None, meeting_id: int | None = None
    ) -> builtins.list[HeadlineListItem]:
        """Get headlines for a user or a meeting.

        Args:
            user_id: The ID of the user (defaults to initialized user_id)
            meeting_id: The ID of the meeting

        Returns:
            A list of HeadlineListItem model instances containing:
            - id
            - title
            - meeting_details
            - owner_details
            - archived
            - created_at
            - closed_at

        Raises:
            ValueError: If both user_id and meeting_id are provided

        Example:
            ```python
            client.headline.list()
            # Returns: [
            #     HeadlineListItem(
            #         id=1,
            #         title='Headline Title',
            #         meeting_details=MeetingInfo(id=1, title='Team Meeting'),
            #         owner_details=OwnerDetails(id=1, name='John Doe'),
            #         archived=False,
            #         created_at='2023-01-01',
            #         closed_at=None
            #     )
            # ]
            ```

        """
        if user_id and meeting_id:
            raise ValueError("Please provide either user_id or meeting_id, not both.")

        if meeting_id:
            response = self._client.get(f"l10/{meeting_id}/headlines")
        else:
            if user_id is None:
                user_id = self.user_id
            response = self._client.get(f"headline/users/{user_id}")

        response.raise_for_status()
        data = response.json()

        return self._transform_headline_list(data)

    def delete(self, headline_id: int) -> None:
        """Delete a headline.

        Args:
            headline_id: The ID of the headline to delete

        """
        response = self._client.delete(f"headline/{headline_id}")
        response.raise_for_status()
